Workspace Editor: Shadow Blocks

chynu opened the pull request:
Can now add/remove shadow blocks in workspace editor. Uses much of the same code as the toolbox editor. 

**EDIT:** Now created a `ShadowController` class for any editor controller which deals with shadow blocks (issue #151). This just moves around code that has already been reviewed/written in previous PR's.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/google/blockly-devtools/149)
<!-- Reviewable:end -->

chynu committed:
Started debugging shadow blocks

chynu committed:
Merge branch 'develop' into shadow

chynu committed:
Rewrote shadow block functions

chynu committed:
Can now create shadow blocks!

chynu committed:
Shadow block isValid check

chynu committed:
Shadow blocks working finally!

chynu committed:
More bug fixes

chynu committed:
Nits

chynu committed:
Starting for workspace

chynu committed:
Refactoring functions into factory utils

chynu committed:
Removed unnecessary function

chynu committed:
Shadow blocks working for WSContents

chynu committed:
PR responses for #146

chynu committed:
Removed unnecessary line of code.

chynu committed:
Rearranged location of shadow block buttons

chynu committed:
MZZerge branch 'develop' into shadow

chynu committed:
Responses for #146

chynu committed:
Merge branch 'shadow' into shadow-wksp

chynu committed:
Merge branch 'develop' into shadow-wksp

svouse commented:
Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/factory_utils.js, line 1335 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA24qGAqkipxx4l3Od:-KqA24qGAqkipxx4l3Oe:b-fnk08j) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/factory_utils.js#L1335)):*
> ```JavaScript
>  * @return {!Array.<!Blockly.Block>} List of shadow blocks from given list.
>  */
> FactoryUtils.getShadowBlocks = function(blockList) {
> ```

I think this is a duplicate (the function is defined earlier in the file)

---

*[src/factory_utils.js, line 1352 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA2LngucySee7fQVl1:-KqA2LnhF6gLWCW-DvKq:bssfam1) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/factory_utils.js#L1352)):*
> ```JavaScript
>  *     currently loaded.
>  */
> FactoryUtils.markShadowBlocks = function(blocks) {
> ```

maybe we could differentiate the names (markShadowBlock vs markShadowBlocks) more?
or, since this is a rather straightforward for loop, should the method exist at all?

---

*[src/controller/toolbox_controller.js, line 661 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA3SlGR7LodR2-7Ege:-KqA3SlGR7LodR2-7Egf:bg5da99) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L661)):*
> ```JavaScript
>    *     does not need warning text.
>    */
>   isValidShadow_(block, isShadow) {
> ```

This function is now in FactoryUtils

---

*[src/controller/workspace_controller.js, line 399 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA4V4QFrRUTRshcxhO:-KqA4V4QFrRUTRshcxhP:b-6oocfm) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/workspace_controller.js#L399)):*
> ```JavaScript
>    * @private
>    */
>   makeShadowishBlocks_(blockId) {
> ```

I'd change the name to concern user-generated shadow blocks specifically, instead of shadowish

---

*[src/controller/workspace_controller.js, line 475 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA6VIr9cmnsiUQndaO:-KqA6VIr9cmnsiUQndaP:b-av0r07) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/workspace_controller.js#L475)):*
> ```JavaScript
>     const project = this.projectController.getProject();
>     blocks.forEach((block) => {
>       if (!project.hasBlockDefinition(block.type)) {
> ```

how does this check identify whether or not something's a standard block?

---

*[src/view/toolbox_editor_view.js, line 592 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA46OBPzaFHJmg0Oub:-KqA46OBPzaFHJmg0Ouc:b-56h09t) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/view/toolbox_editor_view.js#L592)):*
> ```JavaScript
>    *     block (must be rendered).
>    */
>   unmarkShadowBlock(block) {
> ```

should this be in FactoryUtils for consistency?

---

*[src/view/toolbox_editor_view.js, line 606 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA6AXGtG_7dzjWr2I6:-KqA4qxk0hhSce64ywZb:b-69yu81) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/view/toolbox_editor_view.js#L606)):*
> ```JavaScript
>    *     block position.
>    */
>   enableShadowButtons(isShadow, isValid) {
> ```

See comment on workspace_editor_view

---

*[src/view/workspace_editor_view.js, line 311 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA57Q_TtYQTzOQ-IIt:-KqA57Q_TtYQTzOQ-IIu:batrh6c) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/view/workspace_editor_view.js#L311)):*
> ```JavaScript
>    * @param {boolean=} opt_disableAll Whether to hide both buttons entirely.
>    */
>   showAndEnableShadow(ifAdd, ifEnable, opt_disableAll) {
> ```

This should only be defined in one place (same for below functions). Maybe move to FactoryUtils? Or make a class that these two classes inherit from?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 677 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8_yZE-KB4URB8lHA:-KqA8_yZE-KB4URB8lHB:b8hjmsu) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L677)):*
> ```JavaScript
>    * @private
>    */
>   makeShadowishBlocks_(blockId) {
> ```

also in FactoryUtils

---

*[src/controller/toolbox_controller.js, line 831 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8sE8hvR13ZXjFuf_:-KqA8sE8hvR13ZXjFufa:bth9xg3) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L831)):*
> ```JavaScript
>    *     otherwise.
>    */
>   isUserGenShadowBlock(blockId) {
> ```

now in FactoryUtils

---

*[src/controller/toolbox_controller.js, line 842 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8vYu_orH2mPSP0tB:-KqA8vYu_orH2mPSP0tC:b-rm6psa) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L842)):*
> ```JavaScript
>    * shadow blocks in the view but are still editable and movable.
>    */
>   convertShadowBlocks() {
> ```

also in FactoryUtils I think?

---

*[src/controller/toolbox_controller.js, line 894 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA9-9U0475CH6w2jDl:-KqA9-9U0475CH6w2jDm:b-ly78v7) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L894)):*
> ```JavaScript
>    * @private
>    */
>   warnForUndefinedBlocks_() {
> ```

in FactoryUtils

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

chynu committed:
PR responses

chynu commented:
Review status: 0 of 7 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/factory_utils.js, line 1335 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA24qGAqkipxx4l3Od:-KqALpurydHwih5j_Yyy:bw10jmr) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/factory_utils.js#L1335)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

I think this is a duplicate (the function is defined earlier in the file)
</blockquote></details>

Oops, done!

---

*[src/factory_utils.js, line 1352 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA2LngucySee7fQVl1:-KqALr_i8Q5mP6f4E3Tk:b5b79sl) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/factory_utils.js#L1352)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

maybe we could differentiate the names (markShadowBlock vs markShadowBlocks) more?
or, since this is a rather straightforward for loop, should the method exist at all?
</blockquote></details>

true. removed markShadowBlocks().

---

*[src/controller/toolbox_controller.js, line 661 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA3SlGR7LodR2-7Ege:-KqAHRuTOr3INA3bnreF:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L661)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

This function is now in FactoryUtils
</blockquote></details>

Done.

---

*[src/controller/toolbox_controller.js, line 677 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8_yZE-KB4URB8lHA:-KqAHPcC4SN1-weT_Ty5:b-c5lhzq) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L677)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

also in FactoryUtils
</blockquote></details>

I think this one is still in this class?

---

*[src/controller/toolbox_controller.js, line 831 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8sE8hvR13ZXjFuf_:-KqAHYmgl_kUDBh_Ajtc:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L831)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

now in FactoryUtils
</blockquote></details>

Done.

---

*[src/controller/toolbox_controller.js, line 842 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8vYu_orH2mPSP0tB:-KqAHfTTCdIzQEPv3T38:b-gn9tyj) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L842)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

also in FactoryUtils I think?
</blockquote></details>

I think this one is still in the class?

---

*[src/controller/toolbox_controller.js, line 894 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA9-9U0475CH6w2jDl:-KqAIIGbAzgVthJZjOn0:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L894)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

in FactoryUtils
</blockquote></details>

Done.

---

*[src/controller/workspace_controller.js, line 399 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA4V4QFrRUTRshcxhO:-KqAIu6VOoHj1L5GCR7k:b-nul2ar) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/workspace_controller.js#L399)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

I'd change the name to concern user-generated shadow blocks specifically, instead of shadowish
</blockquote></details>

Renamed to `createFakeShadowBlocks_()`.

---

*[src/controller/workspace_controller.js, line 475 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA6VIr9cmnsiUQndaO:-KqAJQZaDD70Gpkx1_Og:bbo58h9) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/workspace_controller.js#L475)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

how does this check identify whether or not something's a standard block?
</blockquote></details>

Ooooh, you're right. I didn't put that in. Done..!

---

*[src/view/toolbox_editor_view.js, line 592 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA46OBPzaFHJmg0Oub:-KqALeXDT_59OZnVBdKN:bm9tzvv) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/view/toolbox_editor_view.js#L592)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

should this be in FactoryUtils for consistency?
</blockquote></details>

You are right. Moved to FactoryUtils.

---

*[src/view/workspace_editor_view.js, line 311 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA57Q_TtYQTzOQ-IIt:-KqAKnyT6DmJSjVgLOBM:b-8y99x5) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/view/workspace_editor_view.js#L311)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

This should only be defined in one place (same for below functions). Maybe move to FactoryUtils? Or make a class that these two classes inherit from?
</blockquote></details>

That is a good idea. I can make this an issue (#151) and tackle it in a future PR, but since this works perhaps we should go through with it for now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

chynu committed:
Created shadow controller class

svouse commented:
Done.

---

Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/model/workspace_contents.js, line 130 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqAQSN_WPOLc7-0O5F8:-KqAQSN_WPOLc7-0O5F9:b-32wr9w) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/model/workspace_contents.js#L130)):*
> ```JavaScript
>    * Determines if a block is a shadow block given a unique block ID.
>    * @param {string} blockId The unique ID of the block to examine.
>    * @return {boolean} True if the block is a user-generated shadow block, false
> ```

Isn't there a FactoryUtils function for this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149#-:-KqAB__d6zH8LzD7bNDZ:b-896fix)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/factory_utils.js, line 1335 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqAzMQC5X1G6m0ow8bj:-KqAzMQC5X1G6m0ow8bk:b32wym) ([raw file](https://github.com/google/blockly-devtools/blob/c8dafcfa5527654f93c20e49336db10db016f3e1/src/factory_utils.js#L1335)):*
> ```JavaScript
>  *     block (must be rendered).
>  */
> FactoryUtils.unmarkShadowBlock = function(block) {
> ```

un

---

*[src/factory_utils.js, line 1345 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqAzNQqzuYCamPaRbCT:-KqAzNQqzuYCamPaRbCU:bb7kj82) ([raw file](https://github.com/google/blockly-devtools/blob/c8dafcfa5527654f93c20e49336db10db016f3e1/src/factory_utils.js#L1345)):*
> ```JavaScript
>  * @param {!Blockly.Block} block Blockly block to be marked as a shadow block.
>  */
> FactoryUtils.markShadowBlock = function(block) {
> ```

mark

---

*[src/controller/toolbox_controller.js, line 677 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8_yZE-KB4URB8lHA:-KqAzZdgS5Ojblezyiub:bm5qjt6) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L677)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think this one is still in this class?
</blockquote></details>

my b- I was thinking of the marShadow and unmarkShadow functions in FactoryUtils

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/factory_utils.js, line 1335 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqAzMQC5X1G6m0ow8bj:-KqOV5oeVKeS4nG60Ka7:bqyx32q) ([raw file](https://github.com/google/blockly-devtools/blob/c8dafcfa5527654f93c20e49336db10db016f3e1/src/factory_utils.js#L1335)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

un
</blockquote></details>

disregard this+ the other comment I accidentally published them//had been using them to track locations

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 8 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 842 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8vYu_orH2mPSP0tB:-KqOWsZiRYOk5iLOI-20:b-hd3x2i) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L842)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

I think this one is still in the class?
</blockquote></details>

My b again- although I suppose I'm now wondering, is there a clear logical division between the shadow functions in FactoryUtils and the one's that aren't?//Could we create one for organization 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/controller/workspace_controller.js, line 349 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOXN5OaWtUROinbi88:-KqOXN5OaWtUROinbi89:b-nqhv5v) ([raw file](https://github.com/google/blockly-devtools/blob/c8dafcfa5527654f93c20e49336db10db016f3e1/src/controller/workspace_controller.js#L349)):*
> ```JavaScript
>    * preview when done.
>    */
>   addShadow() {
> ```

Now thinking about it more, as per above comment, I think there should be a clear location for all shadow block functions

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

chynu committed:
Making new shadow block class for #151

chynu commented:
Review status: 0 of 9 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 842 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8vYu_orH2mPSP0tB:-KqOaPovYMqhqjRCAaY3:b-d6sut7) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L842)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

My b again- although I suppose I'm now wondering, is there a clear logical division between the shadow functions in FactoryUtils and the one's that aren't?//Could we create one for organization 
</blockquote></details>

Yeah it's kind of weird right now.. It will be resolved once I address #151.

---

*[src/controller/workspace_controller.js, line 349 at r2](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOXN5OaWtUROinbi88:-KqOagqXxxXFZth4uv4G:b-q6j60v) ([raw file](https://github.com/google/blockly-devtools/blob/c8dafcfa5527654f93c20e49336db10db016f3e1/src/controller/workspace_controller.js#L349)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Now thinking about it more, as per above comment, I think there should be a clear location for all shadow block functions
</blockquote></details>

Yes, that is a good point. It'll probably all go into the shadow block class as described in #151.

---

*[src/model/workspace_contents.js, line 130 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqAQSN_WPOLc7-0O5F8:-KqOSxF7DRyZXaYr4j8l:bqbayfd) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/model/workspace_contents.js#L130)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

Isn't there a FactoryUtils function for this?
</blockquote></details>

Nope, I don't think so. `isShadowBlock()` here checks the ID against an array of ID's for user-made shadow blocks, whereas `FactoryUtils.isValidShadowBlock()` is checking whether a block does not break shadow block rules (and therefore whether it _can_ be a shadow block, regardless of whether it actually is or isn't).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

chynu commented:
Created new `ShadowController` class. Allows `WorkspaceController` and `ToolboxController` to share shadow block functions.

svouse commented:
Review status: 0 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 842 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA8vYu_orH2mPSP0tB:-KqOsDIjKbPJpiG6RzwE:bb7ohn6) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/controller/toolbox_controller.js#L842)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

Yeah it's kind of weird right now.. It will be resolved once I address #151.
</blockquote></details>

sgtm

---

*[src/view/workspace_editor_view.js, line 311 at r1](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqA57Q_TtYQTzOQ-IIt:-KqOsTh062JBFcOHME1v:bb7ohn6) ([raw file](https://github.com/google/blockly-devtools/blob/974f24a112bd926562817c4dfbcd32d96bf319ee/src/view/workspace_editor_view.js#L311)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

That is a good idea. I can make this an issue (#151) and tackle it in a future PR, but since this works perhaps we should go through with it for now?
</blockquote></details>

sgtm

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/controller/workspace_controller.js, line 28 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOtxJ-T6c5yfRMfqQc:-KqOtxJ0oSOuaASDf7ey:bop66l5) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/controller/workspace_controller.js#L28)):*
> ```JavaScript
> goog.require('WorkspaceConfiguration');
> goog.require('WorkspaceContents');
> goog.require('WorkspaceEditorView');
> ```

goog.require('ShadowController')

---

*[src/view/workspace_editor_view.js, line 105 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOuW6O8CAizEsw2Itf:-KqOuW6O8CAizEsw2Itg:b-mc834c) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/view/workspace_editor_view.js#L105)):*
> ```JavaScript
>     this.selectedBlock = null;
> 
>     /**
> ```

nit /* (/** only applies to the next line)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 25 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOukX-T9RPmKwt63rR:-KqOukX-T9RPmKwt63rS:bop66l5) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/controller/toolbox_controller.js#L25)):*
> ```JavaScript
> goog.provide('ToolboxController');
> 
> goog.require('ToolboxEditorView');
> ```

goog.require('ShadowController')

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 9 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/model/toolbox.js, line 342 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOuxew45Ec-ibt8y1b:-KqOuxew45Ec-ibt8y1c:batcsfq) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/model/toolbox.js#L342)):*
> ```JavaScript
>    *     otherwise.
>    */
>   isShadowBlock(blockId) {
> ```

maybe put in ShadowController?

---

*[src/model/workspace_contents.js, line 133 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOv5EwoWuIbTGme4Eo:-KqOv5EwoWuIbTGme4Ep:b-qg5xrb) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/model/workspace_contents.js#L133)):*
> ```JavaScript
>    *     otherwise.
>    */
>   isShadowBlock(blockId) {
> ```

comment in toolbox.js

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

chynu committed:
Responses to 149

chynu commented:
Review status: 0 of 9 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/controller/toolbox_controller.js, line 25 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOukX-T9RPmKwt63rR:-KqPXPp10rbobZjrqVez:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/controller/toolbox_controller.js#L25)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

goog.require('ShadowController')
</blockquote></details>

Done.

---

*[src/controller/workspace_controller.js, line 28 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOtxJ-T6c5yfRMfqQc:-KqPXUEaluKzkJX-RxS_:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/controller/workspace_controller.js#L28)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

goog.require('ShadowController')
</blockquote></details>

Done.

---

*[src/model/toolbox.js, line 342 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOuxew45Ec-ibt8y1b:-KqPXrm1uNuMNm5SQ8_f:b3sh54g) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/model/toolbox.js#L342)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

maybe put in ShadowController?
</blockquote></details>

This can't go in ShadowController because each Toolbox model object has a unique array of block ID's that are shadow blocks only for that given Toolbox. I think the controller is for marking and generating shadow blocks for import/export/view things related to the application, but there should at least be some sort of storage at the model level of which blocks for a given toolbox are shadow blocks.

---

*[src/view/workspace_editor_view.js, line 105 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOuW6O8CAizEsw2Itf:-KqPYx9-A6QDdpyTsIdB:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/view/workspace_editor_view.js#L105)):*
<details><summary><i>Previously, svouse wrote…</i></summary><blockquote>

nit /* (/** only applies to the next line)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

svouse commented:
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/model/toolbox.js, line 342 at r3](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqOuxew45Ec-ibt8y1b:-KqP_K6YMhGnO2SH-aIn:b32wpx) ([raw file](https://github.com/google/blockly-devtools/blob/6f9da2242b73bf49302a9e2a55134195b6b2bfc9/src/model/toolbox.js#L342)):*
<details><summary><i>Previously, chynu (celine choo) wrote…</i></summary><blockquote>

This can't go in ShadowController because each Toolbox model object has a unique array of block ID's that are shadow blocks only for that given Toolbox. I think the controller is for marking and generating shadow blocks for import/export/view things related to the application, but there should at least be some sort of storage at the model level of which blocks for a given toolbox are shadow blocks.
</blockquote></details>

kk

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
Reviewed 1 of 7 files at r1, 1 of 5 files at r2, 4 of 6 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/factory_utils.js, line 1317 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPhhXUSb2i8IWDia69:-KqPhhXVbkhdQcOvnIfh:b-ma516o) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/factory_utils.js#L1317)):*
> ```JavaScript
> /**
>  * Sets a warning on blocks that are not defined within a given project.
>  * @param {!Array.<Blockly.Block>} blocks A list of blocks to check if they are
> ```

!Array.<!Blockly.Block> The items in the array should also not be null.

---

*[src/controller/shadow_controller.js, line 134 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPy2sv-z9Auirn5Uhr:-KqPy2sv-z9Auirn5Uhs:b-amn25y) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L134)):*
> ```JavaScript
>     // Check if block has variables (variable blocks cannot be shadow blocks).
>     const hasVar = FactoryUtils.hasVariableField(selected);
>     this.view.enableShadowButtons(isShadow, isValid);
> ```

hasVar should be taken into account.

---

*[src/controller/shadow_controller.js, line 266 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPxtCyAKOtLDBQ59GX:-KqPxtCyAKOtLDBQ59GY:btqchj) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L266)):*
> ```JavaScript
>    * @param {!Blockly.Block} block The block to be converted to a user-generated
>    *     shadow block.
>    * @private
> ```

@recursive

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

AnmAtAnm commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[src/controller/shadow_controller.js, line 33 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPxvpueENSCzZfeGjG:-KqPxvpueENSCzZfeGjH:bj36de) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L33)):*
> ```JavaScript
>  * @authors sagev (Sage Vouse), celinechoo (Celine Choo)
>  */
> class ShadowController {
> ```

Not pushing for a change, but I probably would have designed this as an encapsulated helper class, rather than a controller base class. No action necessary.

---

*[src/controller/shadow_controller.js, line 34 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPyFe0wkOgZnBbpEmI:-KqPyFe0wkOgZnBbpEmJ:b-uq20w3) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L34)):*
> ```JavaScript
>  */
> class ShadowController {
>   constructor(projectController, hiddenWorkspace) {
> ```

JsDoc with `@param` documentation.

---

*[src/controller/shadow_controller.js, line 220 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPyadwIakC96Nquzra:-KqPyadwIakC96Nquzrb:bka5vk2) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L220)):*
> ```JavaScript
>    * Given a set of blocks currently loaded, returns all blocks in the workspace
>    * that are user generated shadow blocks.
>    * @param {Array.<!Blockly.Block>} blocks Array of blocks currently loaded.
> ```

Remove the `.` after `Array`. Technically this is allowed, but so is the shorter form.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

chynu committed:
Merge branch 'develop' into shadow-wksp

chynu committed:
Responses to #149

chynu commented:
Review status: 2 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/factory_utils.js, line 1317 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPhhXUSb2i8IWDia69:-KqQ320_9dAb1ykaiA0D:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/factory_utils.js#L1317)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

!Array.<!Blockly.Block> The items in the array should also not be null.
</blockquote></details>

Done.

---

*[src/controller/shadow_controller.js, line 34 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPyFe0wkOgZnBbpEmI:-KqQ0QgLY97bQIB5-du8:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L34)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

JsDoc with `@param` documentation.
</blockquote></details>

Done.

---

*[src/controller/shadow_controller.js, line 134 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPy2sv-z9Auirn5Uhr:-KqQ2eicExFoQEz2jUq4:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L134)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

hasVar should be taken into account.
</blockquote></details>

Done.

---

*[src/controller/shadow_controller.js, line 220 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPyadwIakC96Nquzra:-KqQ2orDT7Tf5eMEpGj7:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L220)):*
<details><summary><i>Previously, AnmAtAnm (Andrew n marshall) wrote…</i></summary><blockquote>

Remove the `.` after `Array`. Technically this is allowed, but so is the shorter form.
</blockquote></details>

Done.

---

*[src/controller/shadow_controller.js, line 266 at r4](https://reviewable.io:443/reviews/google/blockly-devtools/149#-KqPxtCyAKOtLDBQ59GX:-KqQ2tZUcpYqFPp89G1c:b-896fix) ([raw file](https://github.com/google/blockly-devtools/blob/902a7787303404735741bdfe3a6508e16f417686/src/controller/shadow_controller.js#L266)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

@recursive
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149)*
<!-- Sent from Reviewable.io -->

chynu committed:
Merge branch 'develop' into shadow-wksp

AnmAtAnm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 10 files at r5.
Review status: 5 of 10 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149#-:-KqQ46Y19jdbMTSekpS0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rachel-fenichel commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 5 of 10 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly-devtools/149#-:-KqQ4b00YHNokRqkd4W8:bnfp4nl)*
<!-- Sent from Reviewable.io -->
